### ISSUE TITLE
bad --tag should fail a publish

### ISSUE THREAD

**Issue by rvagg** [npm/npm#7144]
I'm not sure how to achieve this properly in the code so this is just going to be an issue, sorry!

Today I wanted to publish a backport fix of a package that was already up to 1.5.0 but the fix needed to go on 1.4.2 _but_ 1.5.0 should remain `@latest` so I did `npm publish --tag 1.4` and it came up with a warning: "Tag must not be a SemVer range" and published anyway, using `@latest` so I had a bit of a mess to clean up.

Ideally, an error like that on the commandline should not result in side-effects shouldn't it? I would have preferred the opportunity to correct my mistake and try again.

**Comment by smikes** [comment]
I discovered the same thing this morning as I was working on https://github.com/npm/npm/pull/7099 ; a validation failure, even when it's an error, does not stop the process.

**Comment by othiym23** [comment]
This is super gross. Bad tags should definitely stop a publish. A patch would be rad, but I understand that you're probably super busy, @rvagg, so I'll see if I can get to this sometime in the next week. (No promises, though -- you're not the only one up to his ass in alligators.)

Thanks for the report!

**Comment by rvagg** [comment]
I started poking around in the code but couldn't figure out the right way to do this because `tag` is just another config option and there doesn't appear to be any hard-failures on config setting so no precedent in the code.

fwiw I did feel bad making this an issue and not a PR.

**Comment by smikes** [comment]
Candidate implementation (incomplete) here: https://github.com/npm/npm/pull/7099

The problem I got stuck on is that checking that dist-tag is not a valid `semver` value happens at different levels, and when validation fails the dist-tag falls back to `latest`, which passes the "invalid semver" check, but that means that the following workflow has nasty, unexpected results:

```
# receive patch for issue in 2.0.1 and 1.2.2
git merge file.patch
npm version 2.02
npm publish  # tag set to latest, good!

git checkout v1.2.2
git merge file.patch
npm version 1.2.3

# I don't want tag set to latest
npm publish --tag 1.2.3
npm WARN Tag must not be a SemVer range
# published successfully with tag `latest` -- oops
```

**Comment by mcollina** [comment]
@rvagg I reported also this back in December, and I also felt ashamed of not fixing it. It is a little bit more tricky than what It should be. As I can only guess, it needs some biggish refactoring of the whole validation step.

**Comment by othiym23** [comment]
This is fixed by 4ef1412d0061239da2b1c4460ed6db37cc9ded27. All credit to @smikes for making it happen!

**Comment by mcollina** [comment]
This is great! Thanks @smikes!
